File: 1197-minimum-knight-moves/1197-minimum-knight-moves.py
class Solution(object):
    def minKnightMoves(self, x, y):
        """
        :type x: int
        :type y: int
        :rtype: int
        """
        def get_neighbors(coord):
            res = []
            row, col = coord
            delta_row = [-2, -2, -1, 1, 2, 2, 1, -1]
            delta_col = [-1, 1, 2, 2, 1, -1, -2, -2]
            for i in range(len(delta_row)):
                r = row + delta_row[i]
                c = col + delta_col[i]
                res.append((r, c))
            return res

        def bfs(start):
            visited = set()
            steps = 0
            queue = deque([start])
            while len(queue) > 0:
                n = len(queue)
                for _ in range(n):
                    node = queue.popleft()
                    if node[0] == y and node[1] == x:
                        return steps
                    for neighbor in get_neighbors(node):
                        r, c = neighbor
                        if neighbor in visited:
                            continue
                        queue.append(neighbor)
                        visited.add(neighbor)
                steps += 1

        return bfs((0, 0))

File: 1197-minimum-knight-moves/1197-minimum-knight-moves.go
func minKnightMoves(x int, y int) int {

    dirs := [][]int{
        {-2,-1},
        {-2,1},
        {-1,2},
        {1,2},
        {2,1},
        {2,-1},
        {1,-2},
        {-1,-2},
    }

    type position struct{ r,c int }
    steps := 0
    startPos := position{r:0,c:0}
    visited := map[position]struct{}{ startPos: struct{}{} }
    q := []position{startPos}
    for len(q) != 0 {
        qSize := len(q)
        for qSize != 0 {
            
            dq := q[0]
            q = q[1:]
            
            cr := dq.r
            cc := dq.c
            if cr == x && cc == y {
                return steps
            }
            
            for _, dir := range dirs {
                r := cr + dir[0]
                c := cc + dir[1]
                rc := position{r: r, c:c}
                if _, ok := visited[rc]; !ok {
                    visited[rc] = struct{}{}
                    q = append(q, rc)
                }
            }
            qSize--
        }
        steps++
    }
    return -1
}
